### PR TITLE
Delete user space

### PR DESCRIPTION
**Pull request by patnorris**
Allows users to "delete" Spaces
As Spaces are NFTs, they are not to be deleted. We hide them instead, such that the Space cannot be retrieved anymore. It's still available as an NFT though (e.g. if users look into their wallet).

To test:
On the My Spaces tab, find the Delete button for each Space. 
Follow the Delete process.
The Deletion should be successful; the Space doesn't show up anymore.
There should not be any side-effects on other Spaces and functionality.

### CODE REVIEW

**Comment by patnorris** [comment]
These changes are available on devDeploy: https://github.com/Bebb-Protocol-and-Apps/PWS/pull/71

**Comment by tiusty** [comment]
After deleting the space it caused the web page to reload which made me have to log in again. Did this not happen to you? The space succesfully deleted though

File: src/PersonalWebSpace_backend/main.mo [hunk 1]
```diff
@@ -353,11 +354,30 @@ shared actor class PersonalWebSpace(custodian: Principal, init : Types.Dip721Non
   };
 
   public shared query ({caller}) func getCallerSpaces() : async [Types.Nft] {
-    let spaces = List.filter(nfts, func(token: Types.Nft) : Bool { token.owner == caller });
+    let spaces = List.filter(nfts, func(token: Types.Nft) : Bool { 
+      switch (token.owner == caller) {
```
Comment (tiusty) [inline src/PersonalWebSpace_backend/main.mo:358]: Do you need a switch case? or does a simple if statement work?

File: src/PersonalWebSpace_backend/main.mo [hunk 2]
```diff
@@ -614,6 +644,57 @@ shared actor class PersonalWebSpace(custodian: Principal, init : Types.Dip721Non
...
   };
 
+  // Delete a Space (owner only)
+  // Spaces are NFTs and thus not deleted, they are hidden such that they may not be retrieved anymore
+  // To keep track of "deleted", i.e. hidden, Spaces we use a HashMap
+  private var deletedSpaces : HashMap.HashMap<Nat, Types.DeletedNftRecord> = HashMap.HashMap(0, Nat.equal, Hash.hash);
+  // Stable version of the deleted Spaces HashMap for cansiter upgrades
+  stable var deletedSpacesStable : [(Nat, Types.DeletedNftRecord)] = [];
+
+  private func hasSpaceBeenDeleted(spaceId : Types.TokenId) : Bool {
```
Comment (tiusty) [inline src/PersonalWebSpace_backend/main.mo:654]: We might want to rename it from deleted to hidden? While to the user it is deleted, it really is hidden?

**Review by tiusty — DISMISSED** [review]

**Comment by patnorris** [comment]
I had implemented it such that window would reload but changed it now to display the delete button as "Deleted" (and the button disabled)

**Review by tiusty — DISMISSED** [review]

**Review by tiusty — APPROVED** [review]

**Pull request merged**